egs/librispeech/asr/simple_v1/bpe_ctc_att_conformer_decode.py

```diff
@@ -6,83 +6,286 @@
+    # The 3 axes are [seq][path][word]
+
+    # Remove epsilons and -1 from word_seqs
+    token_seqs = k2.ragged.remove_values_leq(token_seqs, 0)
+    word_seqs = k2.ragged.remove_values_leq(word_seqs, 0)
+
+    # Remove repeated sequences to avoid redundant computation later.
+    #
+    # Since k2.ragged.unique_sequences will reorder paths within a seq,
+    # `new2old` is a 1-D torch.Tensor mapping from the output path index
+    # to the input path index.
+    # new2old.numel() == unique_word_seqs.num_elements()
+    unique_token_seqs, _, new2old = k2.ragged.unique_sequences(
+        token_seqs, need_num_repeats=False, need_new2old_indexes=True)
+    # Note: unique_token_seqs still has the same axes as token_seqs
+
+    seq_to_path_shape = k2.ragged.get_layer(unique_token_seqs.shape(), 0)
+
+    # path_to_seq_map is a 1-D torch.Tensor.
+    # path_to_seq_map[i] is the seq to which the i-th path
+    # belongs.
+    path_to_seq_map = seq_to_path_shape.row_ids(1)
+
+    # Remove the seq axis.
+    # Now unique_word_seqs has only two axes [path][word]
+    unique_token_seqs = k2.ragged.remove_axis(unique_token_seqs, 0)
+
+    # token_fsas is an FsaVec with axes [path][state][arc]
+    token_fsas = k2.linear_fsa(unique_token_seqs)
+
+    token_fsas_with_epsilon_loops = k2.add_epsilon_self_loops(token_fsas)
+
+    # lats has token IDs as labels and word IDs as aux_labels.
+    # inv_lats has word IDs as labels and token IDs as aux_labels
+    # Do k2.invert to make it compatible to function compute_am_scores
+    inv_lats = k2.invert(lats)
+    inv_lats = k2.arc_sort(inv_lats) # no-op if inv_lats is already arc-sorted
+    am_scores = compute_am_scores(inv_lats, token_fsas_with_epsilon_loops, path_to_seq_map)
+
+    lats = k2.arc_sort(lats)
+    fgram_lm_lats = _intersect_device(lats, token_fsas_with_epsilon_loops, path_to_seq_map, sorted_match_a=True)
+    fgram_lm_lats = k2.top_sort(k2.connect(fgram_lm_lats.to('cpu')).to(lats.device))
+    # am_scores is computed with log_semiring=True
+    # set log_semiring=True here to make fgram_lm_scores comparable to am_scores
+    fgram_tot_scores = fgram_lm_lats.get_tot_scores(use_double_scores=True, log_semiring=True)
+    fgram_lm_scores = fgram_tot_scores - am_scores
+
+
+    # now compute lm scores from transformer decoder
+    token_ids = k2.ragged.to_list(unique_token_seqs)
+    num_seqs = len(token_ids)
+    time_steps = encoder_memory.shape[0]
+    feature_dim = encoder_memory.shape[2]
+    encoder_memory = encoder_memory.expand(time_steps, num_seqs, feature_dim)
+    memory_mask = memory_mask.expand(num_seqs, time_steps)
+
+    # nll: negative log-likelihood
+    nll = model.decoder_nll(encoder_memory, memory_mask, token_ids=token_ids)
+    assert nll.shape[0] == num_seqs
+    decoder_scores = - nll.sum(dim=1)
+    tot_scores = am_scores + fgram_lm_scores + decoder_scores
+    best_seq_idx = new2old[torch.argmax(tot_scores)]
+    best_word_seq = [k2.ragged.to_list(word_seqs)[0][best_seq_idx]]
```

Comment:
Does it work when there are more than 1 sequence, i.e., when batch_size > 1?

Reply:
not work because I am figuring out a way to handle encoder_memory.